src/data/mv_polynomial/monad.lean

```diff
@@ -0,0 +1,248 @@
+/-
+Copyright (c) 2020 Johan Commelin and Robert Y. Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin and Robert Y. Lewis
+-/
+
+import data.mv_polynomial.rename
+
+/-!
+
+# Monad operations on `mv_polynomial`
+
+This file defines two monadic operations on `mv_polynomial`. Given `p : mv_polynomial σ R`,
+
+* `mv_polynomial.bind₁` and `mv_polynomial.join₁` operate on the variable type `σ`.
+* `mv_polynomial.bind₂` and `mv_polynomial.join₂` operate on the coefficient type `R`.
+
+These operations themselves have algebraic structure: `bind₁` and `join₁` are algebra homs and
+`bind₂` and `join₂` are ring homs.
+
+They interact in convenient ways with `mv_polynomial.rename`, `mv_polynomial.map`,
+`mv_polynomial.vars`, and other polynomial operations.
+Indeed, `mv_polynomial.rename` is the "map" operation for the (`bind₁`, `join₁`) pair,
+whereas `mv_polynomial.map` is the "map" operation for the other pair.
+
+## Implementation notes
+
+We add an `is_lawful_monad` instance for the (`bind₁`, `join₁`) pair.
+The second pair cannot be instantiated as a `monad`,
+since it is not a monad in `Type` but in `CommRing` (or rather `CommSemiRing`).
+
+-/
+
+open_locale big_operators
+noncomputable theory
+namespace mv_polynomial
+open finsupp
+
+variables {σ : Type*} {τ : Type*}
+variables {R S T : Type*} [comm_semiring R] [comm_semiring S] [comm_semiring T]
+
+/--
+`bind₁` is the "left hand side" bind operation on `mv_polynomial`, operating on the variable type.
+Given a polynomial `p : mv_polynomial σ R` and a map `f : σ → mv_polynomial τ R` taking variables
+in `p` to polynomials in the variable type `τ`, `bind₁ f p` replaces each variable in `p` with
+its value under `f`, producing a new polynomial in `τ`. The coefficient type remains the same.
+This operation is an algebra hom.
+-/
+def bind₁ (f : σ → mv_polynomial τ R) : mv_polynomial σ R →ₐ[R] mv_polynomial τ R :=
+aeval f
+
+/--
+`bind₂` is the "right hand side" bind operation on `mv_polynomial`, operating on the coefficient type.
+Given a polynomial `p : mv_polynomial σ R` and a map `f : R → mv_polynomial σ S` taking coefficients
+in `p` to polynomials over a new ring `S`, `bind₂ f p` replaces each coefficient in `p` with its
+value under `f`, producing a new polynomial over `S`. The variable type remains the same.
+This operation is a ring hom.
+-/
+def bind₂ (f : R →+* mv_polynomial σ S) : mv_polynomial σ R →+* mv_polynomial σ S :=
+eval₂_hom f X
+
+/--
+`join₁` is the monadic join operation corresponding to `mv_polynomial.bind₁`. Given a polynomial `p`
+with coefficients in `R` whose variables are polynomials in `σ` with coefficients in `R`,
+`join₁ p` collapses `p` to a polynomial with variables in `σ` and coefficients in `R`.
+This operation is an algebra hom.
+-/
+def join₁ : mv_polynomial (mv_polynomial σ R) R →ₐ[R] mv_polynomial σ R :=
+aeval (ring_hom.id _)
```

Comment:
`ring_hom.id` gets coerced to a function here. Is there a reason you didn't directly use `_root_.id`?

Reply:
Good catch. Fixed.